### PR TITLE
Feature/osint 172

### PR DESCRIPTION
**Pull request by R3D1NM**
# [OSINT-172] 시각화 패널 레이아웃 구현
### Modified client file structure
image
- Grouped components into case/data/tool/visualization
- Each of the visualization has its own folder as it was
- Since there will be 2 more timeline graphs, modified timeline.js as a controller of those graphs and added 2 files for other graphs

### Refactored timeline.js
Changed timeline.js into domainTimeline.js and combined those two functions into one

Here's the domainTimeline.js for domain timeline
```
import React, { useEffect, useState } from 'react'
import {
    Chart as ChartJS,
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend,
} from 'chart.js';
import { Line } from 'react-chartjs-2';
import axios from 'axios';

ChartJS.register(
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend
);

const DomainTimeline = (props) => {
    const isDone = props.isDone
    const [datasets, setDatasets] = useState([
        // 기본 데이터
        {
        label: 'No Data',
        fill: false,
        data: [],
          pointRadius: 0, // 포인트를 숨깁니다.
          backgroundColor: 'rgba(0, 0, 0, 0)', // 배경을 투명하게 만듭니다.
          borderColor: 'rgba(0, 0, 0, 0)', // 테두리를 투명하게 만듭니다.
          borderWidth: 0, // 테두리 두께를 0으로 만듭니다.
    },]); // 데이터셋

    useEffect(() => {
        axios.get('/timeline/post').then((response) => {
        const serverData = response.data;
        const data = serverData.post_dicts;
        if (data.length >0){
        const backgroundColors = [
            'rgba(255, 99, 132, 0.2)',
            'rgba(54, 162, 235, 0.2)',
            'rgba(255, 206, 86, 0.2)',
            'rgba(75, 192, 192, 0.2)',
            'rgba(153, 102, 255, 0.2)',
            'rgba(255, 159, 64, 0.2)',
          // 추가적인 색상을 필요한 만큼 추가
        ];
        const newDatasets = [];
        const uniqueLabels = Array.from(new Set(data.map(post => post['domain_name'])));

        data.forEach((post, index) => {
            let label = post['domain_name'];
            let tag = '';
            const xValue = post['regdate'];
            if (post['domain_name']) {
                label = post['domain_name'];
            }
            const backgroundColor = backgroundColors[index % backgroundColors.length];
        
            const postDataset = {
            label,
            fill: false,
            data: [
            {
                x: post['regdate'],
                y: 10,
                tag,
            },
            ],
            pointRadius: 7,
            backgroundColor,
            borderColor: backgroundColor,
            borderWidth: 2,
            };
            newDatasets.push(postDataset);
        });

        console.log(newDatasets);
        // React 상태로 데이터를 설정
        setDatasets(newDatasets);
        } else {
        // 데이터가 없는 경우, 빈 그래프를 보여주도록 datasets 업데이트
        setDatasets([
            {
            label: 'No Data',
            fill: false,
            data: [],
            pointRadius: 0,
            backgroundColor: 'rgba(0, 0, 0, 0)',
            borderColor: 'rgba(0, 0, 0, 0)',
            borderWidth: 0,
            },
        ]);
        }
        })
        .catch((error) => {
        // 서버에서 500 오류가 발생한 경우 처리
            console.error('서버 오류:', error);
        // 사용자에게 오류 메시지 표시 또는 다른 오류 처리 작업 수행
        });
        console.log(datasets);
    }, [isDone]);

    
    const options = {
        maintainAspectRatio: false,
        showLine: false,
        scales: {
        x: {
            grid: {
            display: false,
            },
        },
        y: {
            suggestedMin: 0,
            suggestedMax: 50,
            display: true,
            ticks: {
            display: false,
            },
            grid: {
            display: false,
            },
        },
        },
        plugins: {
        tooltip: {
            callbacks: {
            label: function (context) {
                var label = context.dataset.label || '';
                if (label) {
                label += ': ';
                }
                var tag =
                context.dataset.data[context.dataIndex].tag !== undefined
                    ? context.dataset.data[context.dataIndex].tag
                    : 'undefined';
                // 'y' 값을 tag에서 제외
                label += tag;
                return label;
            },
            },
        },
    },
    };
    
    return (
        <div>
            <Line options={options} data={{datasets}} style={{height:"200px",width:"840px"}}/>
        </div>
    );
}

export default DomainTimeline
```

## 🙏 To Reviewers
🧑‍🤝‍🧑 @D7MeKz @gyomin0315 
- It works the same, as it was some kind of refactoring
- **BUT**, I modified the domain timeline a little, so plz check it out carefully. I am so ready to explain any of those changes :)
- @gyomin0315 When you add another timeline next time, just put it below the domain timeline in timeline.js. I will fix the layout later.

[OSINT-172]: https://bob-transfer.atlassian.net/browse/OSINT-172?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by D7MeKz — APPROVED** [review]
👍

**Comment by gyomin0315** [comment]
Q. I didn't fully understand your ment "When you add another timeline next time, just put it below the domain timeline in timeline.js. I will fix the layout later". If I add another timeline next time, aren't I supposed to put the new timeline in suspectTimeline.js or WholeTimeline.js ?

**Comment by R3D1NM** [comment]
> Q1. I didn't fully understand your ment "When you add another timeline next time, just put it below the domain timeline in timeline.js. I will fix the layout later". If I add another timeline next time, aren't I supposed to put the new timeline in suspectTimeline.js or WholeTimeline.js ?

Oh, I mean put supspectTImeline and WholeTimeline here
```
//timeline.js
import React from 'react'
import DomainTimeline from './domainTimeline'

const Timeline = (props) => {
  const isDone = props.isDone
  return (
    <div>
      {/* Enter Each Timeline Here */}
      <DomainTimeline isDone={isDone}/>
    </div>
  )
}

export default Timeline
```
You are right, make the other two timeline on each js files, and then import them to timeline.js as what I did with DomainTimeline.


> Q2. What is the purpose of const isDone = props.isDone ?

isDone is to handle the tool running state. So if the running is completed, isDone is set to true. I set dependency on useEffect in domainTimelinejs as below.
```
useEffect(() => {
        axios.get('/timeline/post').then((response) => {
        const serverData = response.data;
        const data = serverData.post_dicts;
        if (data.length >0){
          //some of your codes here...
        setDatasets(newDatasets);
        } else {
        //some of your codes here...
        })
        .catch((error) => {
        // 서버에서 500 오류가 발생한 경우 처리
            console.error('서버 오류:', error);
        // 사용자에게 오류 메시지 표시 또는 다른 오류 처리 작업 수행
        });
        console.log(datasets);
    }, [isDone]); //here❗ 
```
 If the value of isDone changes, whole commands in useEffect will be redone, which means, in this case, we get new data from backend. This is to refresh timeline when OSINT tool gathered new data and it was updated in neo4j DB.
In other words, by the value isDone, we can get any new data in neo4j DB to update timeline when OSINT tool running is done.
